File: WorldWindAndroidSDK/WorldWindAndroid/src/gov/nasa/worldwind/pick/PickSupport.java
/* Copyright (C) 2001, 2012 United States Government as represented by
the Administrator of the National Aeronautics and Space Administration.
All Rights Reserved.
*/
package gov.nasa.worldwind.pick;

import android.graphics.Point;
import gov.nasa.worldwind.geom.Position;
import gov.nasa.worldwind.layers.Layer;
import gov.nasa.worldwind.render.DrawContext;

import java.util.*;

/**
 * @author tag
 * @version $Id$
 */
public class PickSupport
{
    protected Map<Integer, PickedObject> pickableObjects = new HashMap<Integer, PickedObject>();

    public PickSupport()
    {
    }

    public void addPickableObject(PickedObject po)
    {
        this.getPickableObjects().put(po.getColorCode(), po);
    }

    public void addPickableObject(int colorCode, Object o)
    {
        this.getPickableObjects().put(colorCode, new PickedObject(colorCode, o));
    }

    public void addPickableObject(int colorCode, Object o, Position position)
    {
        this.getPickableObjects().put(colorCode, new PickedObject(colorCode, o, position, false));
    }

    public void addPickableObject(int colorCode, Object o, Position position, boolean isTerrain)
    {
        this.getPickableObjects().put(colorCode, new PickedObject(colorCode, o, position, isTerrain));
    }

    public void clearPickList()
    {
        this.getPickableObjects().clear();
    }

    protected Map<Integer, PickedObject> getPickableObjects()
    {
        return this.pickableObjects;
    }

    public PickedObject getTopObject(DrawContext dc, Point pickPoint)
    {
        if (this.getPickableObjects().isEmpty())
            return null;

        int colorCode = dc.getPickColor(pickPoint);
        if (colorCode == 0) // getPickColor returns 0 if the pick point selects the clear color.
            return null;

        PickedObject pickedObject = getPickableObjects().get(colorCode);
        if (pickedObject == null)
            return null;

        return pickedObject;
    }

    public PickedObject resolvePick(DrawContext dc, Point pickPoint, Layer layer)
    {
        PickedObject pickedObject = this.getTopObject(dc, pickPoint);
        if (pickedObject != null)
        {
            if (layer != null)
                pickedObject.setParentLayer(layer);

            dc.addPickedObject(pickedObject);
        }

        this.clearPickList();

        return pickedObject;
    }
}

File: WebWorldWind/src/pick/PickSupport.js
/*
 * Copyright (C) 2014 United States Government as represented by the Administrator of the
 * National Aeronautics and Space Administration. All Rights Reserved.
 */
/**
 * @exports PickSupport
 * @version $Id$
 */
define([],
    function () {
        "use strict";

        /**
         * Constructs a pick-support object.
         * @alias PickSupport
         * @constructor
         * @classdesc Provides methods to assist in picking.
         */
        var PickSupport = function () {
            this.pickableObjects = {};
        };

        /**
         * Adds an object to this list.
         * @param {PickedObject} pickableObject The object to add.
         */
        PickSupport.prototype.addPickableObject = function (pickableObject) {
            this.pickableObjects[pickableObject.color.toByteString()] = pickableObject;
        };

        // Internal. Intentionally not documented.
        PickSupport.prototype.topObject = function (dc, pickPoint) {
            if (Object.keys(this.pickableObjects).length === 0) {
                return null;
            }

            var color = dc.readPickColor(pickPoint);
            if (!color) { // getPickColor returns null if the pick point selects the clear color
                return null;
            }

            return this.pickableObjects[color.toByteString()];
        };

        /**
         * Returns the top-most object in this instance's list of pickable objects as determined be reading the
         * pick color from the frame buffer and identifying the pickable object associated with that color.
         * This method should be called after all picked objects have been added to this pick-support instance.
         * This method clears the current list of pickable objects.
         * @param {DrawContext} dc The current draw context.
         * @returns {PickedObject} The top-most picked object, or null if none is found.
         */
        PickSupport.prototype.resolvePick = function (dc) {
            var pickedObject = this.topObject(dc, dc.pickPoint);

            if (pickedObject) {
                dc.addPickedObject(pickedObject);
            }

            this.pickableObjects = []; // clear the pick list to avoid dangling references

            return pickedObject;
        };

        return PickSupport;
    });